### DISCUSSION TITLE
HIP-CPU as back-end instead of OpenMP

### DISCUSSION THREAD

**Discussion by MathiasMagnus** [AdaptiveCpp/AdaptiveCpp#414] [Ideas]
The recently released [HIP-CPU](https://github.com/ROCm-Developer-Tools/HIP-CPU) project is a pure C++17 implementation of nearly the entire HIP API (eg. cross-lane warp functions included) and it's a header-only implementation (runtime state are all inline global variables) so there are no link requirements either. It's super easy to use and works across OSes.

Could hipSYCL use HIP-CPU as a back-end instead of OpenMP? That would be a possible way forward to see hipSYCL function on Windows for example.

**Comment by illuhad** [comment]
I'm not very familiar with the Windows world. Is OpenMP an issue on windows? Recent hipSYCL does not have any dependencies on LLVM when building only with the CPU backend enabled, so it should be easy to install (but due to lack of testing on Windows there may be hiccups because of path conventions etc).

I'm skeptical about HIP-CPU for a variety of reasons. We have tried building a library implementation of HIP already two years ago (https://github.com/illuhad/hipCPU). HIP/CUDA models were never meant to be implemented in a library-only way, and as it turns out you have to decide between either functionality or performance caveats when attempting this. While hipCPU is probably more simplistic than HIP-CPU, I expect that fundamentally HIP-CPU will have similar limitations as hipCPU had.

The bottom line is that HIP is far lower-level model than SYCL (with the exception of nd_range parallel for). As a consequence this means that building on top of a CPU HIP implementation means that we give the backend less freedom to do the right thing for CPUs.

Additionally, introducing another HIP API is unattractive as it would lead to symbol clashes when targeting both AMD GPUs and CPUs.

Now, if OpenMP is a problem, the good news is I do however have some plans regarding additional CPU backends, most notably regarding TBB and other task graph based libraries. So, additional CPU backends are likely to appear :)

**Reply by fodinabor** [reply]
I do think you can use Clang and GCC on Windows to get recent OpenMP support, just MSVC does only support OpenMP up to version 2.0, which I _think_ is probably not sufficient for hipSYCL?
a library only solution like TBB or similar would thus be more easily supported with the standard MSVC compiler..

**Reply by illuhad** [reply]
Ah, I see. I'd need to check what was introduced in OpenMP 3.0 but actually we don't do a lot of fancy things. From the top of my head we need `parallel, for, collapse, simd, threadprivate` clauses.

**Reply by MathiasMagnus** [reply]
Would benchmarks of HIP-CPU influence the revisit of this decision? It wouldn't take much to run rocPRIM, rocThrust, rocRAND benchmarks which likely have equivalent SYCL implementations (although I'm gravely missing SYCL-PRIM as a project of reusable algorithms).

**Reply by illuhad** [reply]
Are you saying that HIP-CPU is substantially faster than the current CPU backend? If so, since both are library-only implementations I think the goal should be to find out what they are doing differently and feed it into the current CPU backend and future CPU backends in order to resolve any performance issues rather than abandoning it and moving to something new.

I still see little reason to funnel SYCL through HIP on CPU when we could just as well invest that effort into building a TBB backend that would give us far more control (especially when it comes to `single_task` task graph-style computation which HIP cannot do at all).
After all, part of the redesign with the new runtime was precisely to get away from HIP on CPU in order to be able to do this in the future, so reintroducing it seems unattractive.

With the current architecture in hipSYCL, the HIP API is no more important than any other backend API, so there's no reason to prefer HIP over something else, in particular over something that has been designed for CPUs from the start such as TBB.
In fact, it would make deploying and building hipSYCL more difficult as users would have to decide at hipSYCL build time whether they *either* want to use HIP-CPU *or* regular HIP. This is directly opposed to work that we have done to be able to target both CPU and GPU simultaneously.

EDIT: HIP-CPU just uses parallel STL under the hood, so we could just as well go to PSTL directly. It's not clear to me why we would want an additional wrapper like HIP-CPU around it?
